### ISSUE TITLE
Stochasticity in M3GNet prediction

### ISSUE THREAD

**Issue by BowenD-UCB** [materialsvirtuallab/matgl#116]
Hi Shyueping and Kenko

The dgl implentation of M3GNet gives stochastic predictions, while this was not observed with the original M3GNet implementation.

I'm worrying this might associated to some serious bug that needs attention to.
Please have a look:
<img width="1117" alt="Screen Shot 2023-07-26 at 6 27 54 PM" src="https://github.com/materialsvirtuallab/matgl/assets/84425382/6aa381db-536e-4343-98c9-f9fb8c7c6d28">

**Comment by shyuep** [comment]
Hmm interesting. The MEGNet implementation doesn't suffer from this problem. So it must be in the threebody computations. @kenko911 pls fix this asap.

**Comment by shyuep** [comment]
I debugged a little. I can confirm that in the forward method, the node_feat and edge_feat are consistent. But the 
```
node_vec = self.readout(g)
vec = torch.hstack([node_vec, state_feat]) if self.inc...```

are stochastic. So there must be some issue with the readout layer. MEGNet does not use the same readout.

**Comment by kenko911** [comment]
@BowenD-UCB Thanks a lot for reporting the issue and I will figure out the problem and fix it ASAP.

**Comment by JiQi535** [comment]
It seems that M3GNet UP is not having this issue.
<img width="991" alt="Screenshot 2023-07-26 at 7 40 21 PM" src="https://github.com/materialsvirtuallab/matgl/assets/54908836/e466c789-6a7a-4258-860f-3cc31eadb764">

**Comment by shyuep** [comment]
Because the UP is not intensive and so does not call the readout.

**Comment by shyuep** [comment]
I have confirmed that the stochasticity resulted from the Set2Set in dgl.nn. For instance, I replaced the set2set call in Set2SetReadout by the following code:

```
            for i in range(3):
                set2set = Set2Set(in_feats, n_iters=self.num_steps, n_layers=self.num_layers)
                print(g.ndata["node_feat"])
                out_tensor = set2set(g, g.ndata["node_feat"])
                print(f"out={out_tensor}")
```

The out tensor is different all three times, despite there being no change in initialization and the input.

**Comment by shyuep** [comment]
I just pushed a fix. The trick is to init the Set2Set only once. Every time it is init, it introduces stochasticity.

**Comment by shyuep** [comment]
However, it should be noted that multiple loads of the model will still lead to stochasticity. I am not sure how to solve that. I am reopening this issue for now.

**Comment by shyuep** [comment]
@kenko911 I think you need to look into whether there is some random seed somewhere that needs to be fixed for dgl.nn for the Set2Set initialization. It is rather strange it is not deterministic.

**Comment by shyuep** [comment]
I found the actual solution, but not sure what is the best approach. @BowenD-UCB @kenko911 if we set `dgl.seed(42)` (or whatever int), then the answer is completely deterministic. But I am not sure that we should be hard-coding a random seed into matgl. It seems the LSTM process in Set2Set is inherently stochastic. I am happy to hear recommendations.

**Comment by shyuep** [comment]
Scratch what I wrote previously. The dgl.seed does not do anything. Right now, the stochasticity with single model loading is solved. But the stochasticity with multiple model loadings is not.

```
from pymatgen.core import Lattice, Structure
import matgl
import torch
torch.manual_seed(0)
import random
random.seed(0)
import dgl
dgl.seed(0)
# This is the structure obtained from the Materials Project.
struct = Structure.from_spacegroup("Pm-3m", Lattice.cubic(4.1437), ["Cs", "Cl"], [[0, 0, 0], [0.5, 0.5, 0.5]])
for i in range(10):
    model = matgl.load_model("M3GNet-MP-2018.6.1-Eform")
    eform = model.predict_structure(struct)
    print(f"The predicted formation energy for CsCl is {float(eform.numpy()):.3f} eV/atom.")
```

still gives stochastic results.

**Comment by shyuep** [comment]
Note that it seems that if you fix the torch.manual_seed, all stochasticity is gone.

This is the fix that I pushed. I am still not sure why this non-deterministic behavior occurs in m3gnet but not MEGNet. MEGnet seems to work fine without any manual seed set. It may be due to the number of layers in the set2set? M3Gnet uses 3 layers while MEGnet uses just 1.

**Comment by shyuep** [comment]
@BowenD-UCB Just a note that Kenko pushed a final fix this morning. A new version has been released. The main issue was that the final layer was not in the init and so the weights were not saved. The new code fixes this problem. The only model that was affected was the M3GNet formation energy model. I also added an integration test to make sure it does not recur in future.
